src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs

```diff
@@ -298,11 +332,222 @@ void writeEnumElement(string kind, int value, string operationName, string? extr
+                    Write(": base(");
+
+                    var @base = _typeMap[type.Base];
+                    if (@base is object)
+                    {
+                        foreach (var prop in GetAllProperties(@base).Where(p => !IsIOperationType(p.Type)))
+                        {
+                            Write($"{prop.Name.ToCamelCase()}, ");
+                        }
+                    }
+
+                    writeStandardBaseArguments(type switch
+                    {
+                        { IsAbstract: true } => "kind",
+                        { IsInternal: true } => "OperationKind.None",
+                        _ => $"OperationKind.{getKind(type)}"
+                    });
+                    Outdent();
+                    // The base class is responsible for initializing its own properties,
+                    // so we only initialize our own properties here.
+                    if (type.Properties.Count == 0)
+                    {
+                        // Note: our formatting style is a space here
+                        WriteLine(" { }");
+                    }
+                    else
+                    {
+                        Blank();
+                        Brace();
+                        foreach (var prop in type.Properties)
+                        {
+                            WriteLine($"{prop.Name} = {prop.Name.ToCamelCase()};");
+                        }
+                        Unbrace();
+                    }
+
+                    foreach (var prop in type.Properties)
+                    {
+                        WriteLine($"public {prop.Type} {prop.Name} {{ get; }}");
+                    }
+
+                    if (type is Node node)
+                    {
+
+                        var visitorName = GetVisitorName(node);
+                        WriteLine($@"public override IEnumerable<IOperation> Children => Array.Empty<IOperation>();
+        public override void Accept(OperationVisitor visitor) => visitor.{visitorName}(this);
+        public override TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument) => visitor.{visitorName}(this, argument);");
+                    }
+
+                    Unbrace();
+                }
+            }
+
+            WriteLine("#endregion");
+
+            void writeStandardConstructorParameters(bool isAbstract)
+            {
+                if (isAbstract)
+                {
+                    Write("OperationKind kind, ");
+                }
+
+                WriteLine("SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit)");
+            }
+
+            void writeStandardBaseArguments(string kind)
+            {
+                Write($"{kind}, semanticModel, syntax, type, constantValue, isImplicit)");
+            }
 
-        private bool IsImmutableArray(string typeName)
+            string getKind(AbstractNode node)
```

Comment:
>getKind [](start = 19, length = 7)

static?

Reply:
It uses `_typeMap`, can't be.

---
In reply to: [314004288](https://github.com/dotnet/roslyn/pull/37791#discussion_r314004288) [](ancestors = 314004288)